ext/mbstring/mbstring.c

```diff
@@ -2838,37 +2838,258 @@ PHP_FUNCTION(mb_decode_mimeheader)
+			if (c >= 0xFF61 && c <= 0xFF9F) {
+				int n = c - 0xFF60;
+				if (next >= 0xFF61 && next <= 0xFF9F) {
+					if (next == 0xFF9E && ((n >= 22 && n <= 36) || (n >= 42 && n <= 46))) {
+						*consumed = true;
+						return 0x3001 + hankana2zenkana_table[n];
+					} else if (next == 0xFF9E && n == 19) {
+						*consumed = true;
+						return 0x30F4;
+					} else if (next == 0xFF9F && n >= 42 && n <= 46) {
+						*consumed = true;
+						return 0x3002 + hankana2zenkana_table[n];
+					}
+				}
+
+				return 0x3000 + hankana2zenkana_table[n];
+			}
+		} else if ((mode & MBFL_HAN2ZEN_HIRAGANA) && (mode & MBFL_HAN2ZEN_GLUE)) {
+			if (c >= 0xFF61 && c <= 0xFF9F) {
+				int n = c - 0xFF60;
+				if (next >= 0xFF61 && next <= 0xFF9F) {
+					if (next == 0xFF9E && ((n >= 22 && n <= 36) || (n >= 42 && n <= 46))) {
+						*consumed = true;
+						return 0x3001 + hankana2zenhira_table[n];
+					} else if (next == 0xFF9F && n >= 42 && n <= 46) {
+						*consumed = true;
+						return 0x3002 + hankana2zenhira_table[n];
+					}
+				}
+
+				return 0x3000 + hankana2zenhira_table[n];
+			}
+		} else if ((mode & MBFL_HAN2ZEN_KATAKANA) && c >= 0xFF61 && c <= 0xFF9F) {
+			return 0x3000 + hankana2zenkana_table[c - 0xFF60];
+		} else if ((mode & MBFL_HAN2ZEN_HIRAGANA) && c >= 0xFF61 && c <= 0xFF9F) {
+			return 0x3000 + hankana2zenhira_table[c - 0xFF60];
+		}
+	}
+
+	if (mode & MBFL_HAN2ZEN_SPECIAL) { /* special ascii to symbol */
+		if (c == '\\' || c == 0xA5) { /* YEN SIGN */
+			return 0xFFE5; /* FULLWIDTH YEN SIGN */
+		} else if (c == 0x7E || c == 0x203E) {
+			return 0xFFE3; /* FULLWIDTH MACRON */
+		} else if (c == '\'') {
+			return 0x2019; /* RIGHT SINGLE QUOTATION MARK */
+		} else if (c == '"') {
+			return 0x201D; /* RIGHT DOUBLE QUOTATION MARK */
+		}
+	}
+
+	if (mode & (MBFL_ZEN2HAN_ALL | MBFL_ZEN2HAN_ALPHA | MBFL_ZEN2HAN_NUMERIC | MBFL_ZEN2HAN_SPACE)) {
+		/* Zenkaku to Hankaku */
+		if ((mode & MBFL_ZEN2HAN_ALL) && c >= 0xFF01 && c <= 0xFF5D && c != 0xFF02 && c != 0xFF07 && c != 0xFF3C) {
+			/* all except " ' \ ~ */
+			return c - 0xFEE0;
+		} else if ((mode & MBFL_ZEN2HAN_ALPHA) && ((c >= 0xFF21 && c <= 0xFF3A) || (c >= 0xFF41 && c <= 0xFF5A))) {
+			return c - 0xFEE0;
+		} else if ((mode & MBFL_ZEN2HAN_NUMERIC) && (c >= 0xFF10 && c <= 0xFF19)) {
+			return c - 0xFEE0;
+		} else if ((mode & MBFL_ZEN2HAN_SPACE) && (c == 0x3000)) {
+			return ' ';
+		} else if ((mode & MBFL_ZEN2HAN_ALL) && (c == 0x2212)) { /* MINUS SIGN */
+			return '-';
+		}
+	}
+
+	if (mode & (MBFL_ZEN2HAN_KATAKANA | MBFL_ZEN2HAN_HIRAGANA)) {
+		/* Zenkaku kana to hankaku kana */
+		if ((mode & MBFL_ZEN2HAN_KATAKANA) && c >= 0x30A1 && c <= 0x30F4) {
+			/* Zenkaku katakana to hankaku kana */
+			int n = c - 0x30A1;
+			if (zenkana2hankana_table[n][1]) {
+				*second = 0xFF00 + zenkana2hankana_table[n][1];
+			}
+			return 0xFF00 + zenkana2hankana_table[n][0];
+		} else if ((mode & MBFL_ZEN2HAN_HIRAGANA) && c >= 0x3041 && c <= 0x3093) {
```

Comment:
Should probably drop the `} else ` here and in the lines below as well.

Reply:
Ack.